source/common/http/codes.cc

```diff
@@ -7,124 +7,147 @@
         .inc();
     info.cluster_scope_
-        .counter(fmt::format("{}zone.{}.{}.upstream_rq_{}", info.prefix_, info.from_zone_,
-                             info.to_zone_, group_string))
+        .counter(join({prefix_without_trailing_dot, zone_, info.from_zone_, info.to_zone_,
+                       absl::StrCat(upstream_rq_, group_string)}))
         .inc();
     info.cluster_scope_
-        .counter(fmt::format("{}zone.{}.{}.upstream_rq_{}", info.prefix_, info.from_zone_,
-                             info.to_zone_, response_code))
+        .counter(join({prefix_without_trailing_dot, zone_, info.from_zone_, info.to_zone_,
+                       absl::StrCat(upstream_rq_, response_code)}))
         .inc();
   }
 }
 
-void CodeUtility::chargeResponseTiming(const ResponseTimingInfo& info) {
-  info.cluster_scope_.histogram(info.prefix_ + "upstream_rq_time")
+void CodeStatsImpl::chargeResponseTiming(const ResponseTimingInfo& info) const {
+  info.cluster_scope_.histogram(absl::StrCat(info.prefix_, upstream_rq_time_))
       .recordValue(info.response_time_.count());
   if (info.upstream_canary_) {
-    info.cluster_scope_.histogram(info.prefix_ + "canary.upstream_rq_time")
+    info.cluster_scope_.histogram(absl::StrCat(info.prefix_, canary_upstream_rq_time_))
         .recordValue(info.response_time_.count());
   }
 
   if (info.internal_request_) {
-    info.cluster_scope_.histogram(info.prefix_ + "internal.upstream_rq_time")
+    info.cluster_scope_.histogram(absl::StrCat(info.prefix_, internal_upstream_rq_time_))
         .recordValue(info.response_time_.count());
   } else {
-    info.cluster_scope_.histogram(info.prefix_ + "external.upstream_rq_time")
+    info.cluster_scope_.histogram(absl::StrCat(info.prefix_, external_upstream_rq_time_))
         .recordValue(info.response_time_.count());
   }
 
   if (!info.request_vcluster_name_.empty()) {
     info.global_scope_
-        .histogram("vhost." + info.request_vhost_name_ + ".vcluster." +
-                   info.request_vcluster_name_ + ".upstream_rq_time")
+        .histogram(join({vhost_, info.request_vhost_name_, vcluster_, info.request_vcluster_name_,
+                         upstream_rq_time_}))
         .recordValue(info.response_time_.count());
   }
 
   // Handle per zone stats.
   if (!info.from_zone_.empty() && !info.to_zone_.empty()) {
     info.cluster_scope_
-        .histogram(fmt::format("{}zone.{}.{}.upstream_rq_time", info.prefix_, info.from_zone_,
-                               info.to_zone_))
+        .histogram(join({stripTrailingDot(info.prefix_), zone_, info.from_zone_, info.to_zone_,
+                         upstream_rq_time_}))
         .recordValue(info.response_time_.count());
   }
 }
 
+absl::string_view CodeStatsImpl::stripTrailingDot(absl::string_view str) {
+  if (absl::EndsWith(str, ".")) {
+    str.remove_suffix(1);
+  }
+  return str;
+}
+
+std::string CodeStatsImpl::join(const std::vector<absl::string_view>& v) {
```

Comment:
Could you probably use this function here?
https://github.com/envoyproxy/envoy/blob/9ba0a437eb23adc4ee3c5fdaeee53e877ea94f7c/source/common/common/utility.h#L309